### PR TITLE
[SYCL] Correctly free the memory allocated in a sub-device context

### PR DESCRIPTION
**Pull request by smaslov-intel**
Signed-off-by: Sergey V Maslov <sergey.v.maslov@intel.com>

### CODE REVIEW

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 1]
```diff
@@ -1519,9 +1519,16 @@ pi_device _pi_platform::getDeviceFromNativeHandle(ze_device_handle_t ZeDevice) {
...
+  // handle + Level-Zero compute group/engine index], so there is now no 1:1
+  // mapping from L0 device handle to PI device assumed in this function. Until
+  // Level-Zero adds unique ze_device_handle_t for sub-sub-devices, here we
+  // filter out PI sub-sub-devices.
   auto it = std::find_if(PiDevicesCache.begin(), PiDevicesCache.end(),
                          [&](std::unique_ptr<_pi_device> &D) {
-                           return D.get()->ZeDevice == ZeDevice;
+                           return D.get()->ZeDevice == ZeDevice &&
+                                  (D.get()->RootDevice == nullptr ||
+                                   D.get()->RootDevice->RootDevice == nullptr);
```
Comment (bso-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:1531]: It looks to me that you are filtering out both sub-devices and sub-sub-devices.
This means you are going to return only root devices for the given ZeDevice handle.
Is that what you wanted?

File: sycl/plugins/level_zero/pi_level_zero.cpp [hunk 2]
```diff
@@ -6267,11 +6274,16 @@ static pi_result USMFreeHelper(pi_context Context, void *Ptr) {
...
+    pi_device Device;
+    if (Context->Devices.size() == 1) {
+      Device = Context->Devices[0];
+      PI_ASSERT(Device->ZeDevice == ZeDeviceHandle, PI_INVALID_DEVICE);
+    } else {
+      // All devices in the context are of the same platform.
+      auto Platform = Context->Devices[0]->Platform;
+      Device = Platform->getDeviceFromNativeHandle(ZeDeviceHandle);
+      PI_ASSERT(Device, PI_INVALID_DEVICE);
+    }
```
Comment (bso-intel) [inline sycl/plugins/level_zero/pi_level_zero.cpp:6286]: Could you explain why this change solved the memory free problem in the sub-device context?

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1531]: This code *allows* root-devices and sub-devices, thus filtering out sub-sub-devices *only*.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 2 above)
Comment (smaslov-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6286]: I cases where context has exactly one PI device(most of the times), we don't need to search for it in the cache. So this code avoids calling "broken" search, if it can be avoided.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 1 above)
Comment (bso-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:1531]: Thanks.

File: sycl/plugins/level_zero/pi_level_zero.cpp (hunk 2 above)
Comment (bso-intel) [reply sycl/plugins/level_zero/pi_level_zero.cpp:6286]: Thanks.

**Review by bso-intel — APPROVED** [review]
LGTM
